keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -856,11 +864,25 @@ termination in Beta. When Pods with sidecars are terminated:
 - Sidecars should terminate serially and in reverse order. I.e. the first
   sidecar to initialize should be the last sidecar to terminate.
 
-To address these needs, before promoting this enhancement to Beta, we introduce
-additional termination grace period fields to manage termination duration
-([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
-and leverage these fields to add reverse order termination of sidecar containers
-after the primary containers terminate.
+1. Using `SIGTERM` signal to notify a sidecar container when regular containers have terminated:
+   - this is the same signal that is used to notify regular containers that the Pod is terminating
+   - it is sent after the previous container has terminated to guarantee ordering
+   - advantages:
+     - it is simple and consistent with regular containers
+     - it should not break existing sidecars implemented as regular containers
```

Comment:
FYI:
What I was trying to say in the last meeting was that
preStop hook and emitting SIGTERM are not completely separate concepts (at least in terms of their implementation).

In my POV, they are one task(container termination) and the task has consecutive works(executing the preStop hook and emitting SIGTERM) sharing the deadline(terminationGracePeriodSeconds).

My point is that it may not be consistent or easy to implement.
(just want to point out its complexity)

Reply:
Thanks for pointing this out @gjkim42 so what you're saying it would be easier to initiate the whole `killContainer` (hook + sigterm) for a sidecar once the previous container (or sidecar) has terminated, rather than separating their concerns?
Sounds reasonable if that is possible at all. WDYT @SergeyKanzhelev @tzneal ?

Reply:
> what you're saying it would be easier to initiate the whole `killContainer` (hook + sigterm) for a sidecar once the previous container (or sidecar) has terminated, rather than separating their concerns?

Yes, I prefer that way.

If the goal is to implement termination ordering while also providing a termination graceful period for sidecar containers without introducing any new field. I'm okay with the current way(separate prestop and SIGTERM). However, it may not be as simple as we thought.

Reply:
Yes, it could work.

From the user's perspective, I think of a `preStop` as a part of the container termination, so separating it looks a bit complicated. It looks quite exceptional.

---
> - Sidecars should not begin termination until all primary containers have terminated.
>   - Implicit in this is that sidecars should continue to be restarted until all primary containers have terminated.

One other question is what happens if the sidecar container crashes during the regular container's termination?
If we want to ignore the crashed sidecar container, it is just ok.
else if we want to restart the sidecar container, how can we handle the preStop hook?

Reply:
> I may be misunderstanding, but your suggestion is that pre-stop and sigterm are essentially linked events, and a sidecar wouldn't get the pre-stop until all main containers and all sidecars that started after it have stopped?

Yes.

My point was that if we choose early pre-stop, even after executing the pre-stop we may have to restart the sidecar container (because it crashes after executing the pre-stop hook). It looks weird to me.

Reply:
> My point was that if we choose early pre-stop, even after executing the pre-stop we may have to restart the sidecar container (because it can crash after executing the pre-stop hook). It looks weird to me.

Reply:
> The advantage to the early pre-stop is that for the case of main containers that consume the full grace period, sidecars are at least aware that they'll be shutting down for a period of time before they just get a SIGKILL when the full grace period is expired.

Yes, I agree that we have to provide some way for a sidecar to be aware of its termination.

However, it would be better not to use pre-stop but to add or borrow the dedicated termination period for the sidecars to make them aware of its termination. (And yes it may need an extra field but I think it is necessary.)

This might just be my own opinion, so I'd appreciate hearing more thoughts.

Reply:
> getting zero notice before a SIGKILL if the main containers don't behave well.

https://github.com/kubernetes/kubernetes/blob/58ce7342230a943b637ff3d129eb3774d5a68cbb/pkg/kubelet/kuberuntime/kuberuntime_container.go#L741-L744

I know it is not enough but we have assigned `minimumGracePeriod` (2s).